### DISCUSSION TITLE
Matrix.os fails to match

### DISCUSSION THREAD

**Discussion by arky** [community/community#25986] [Actions]
<p>Am trying to build pyinstaller executables for multiple OSs, Since MS windows uses “;” separators I used ‘matrix.os’ to switch the run command.</p>
<p>However ‘build with pyinstaller on gnu/Linux and MacOS’ step runs on all OS including MS windows.</p>
<p>Could someone point what I am missing here. Thanks</p>
<pre><code class="lang-auto">name: Package Excalibur with Pyinstaller

on:
  push:
    branches: [master]
  pull_request:
    branches: [master]

jobs:
  build:
    runs-on: ${{ matrix.config.os }}

    strategy:
      matrix:
        config:
          - { os: macos-latest, arch: x64, name: macos-latest }
          - { os: ubuntu-latest, arch: x64, name: ubuntu-latest }
          - { os: windows-latest, arch: x64, name: windows-latest }
          - { os: windows-latest, arch: ia32, name: windows-latest-32 }

    steps:
      - uses: actions/checkout@v1
      - name: Set up Python 3.8
        uses: actions/setup-python@v1
        with:
          python-version: 3.8
      - name: Install dependencies
        run: |
          python -m pip install --upgrade pip
          pip install pyinstaller
      - name: build with pyinstaller on gnu/Linux and MacOS
        if: matrix.os == 'ubuntu-latest' || 'macos-latest'
        run: |
          pyi-makespec --paths=excalibur/executors/celery_executor.py arthur.py
          pyinstaller --onefile --add-data "excalibur/www/templates:excalibur/www/templates" --add-data "excalibur/www/static:excalibur/www/static" --add-data "excalibur/config_templates:excalibur/config_templates" arthur.py
      - name: build with pyinstaller on MS Windows
        if: matrix.os == 'windows-latest'
        run: |
          pyi-makespec --paths=excalibur/executors/celery_executor.py arthur.py
          pyinstaller --onefile --add-data "excalibur/www/templates;excalibur/www/templates" --add-data "excalibur/www/static;excalibur/www/static" --add-data "excalibur/config_templates;excalibur/config_templates" arthur.py
  #  - uses: actions/upload-artifact@v2
  #     with:
  #       name: name-of-artifact
  #       path: src/dist/
</code></pre>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="arky" data-post="1" data-topic="121009">
<div class="title">
<div class="quote-controls"></div>
 arky:</div>
<blockquote>
<p><code>        if: matrix.os == 'ubuntu-latest' || 'macos-latest'</code></p>
</blockquote>
</aside>
<p>I’d guess the problem is the missing <code>matrix.os == </code> before <code>'macos-latest'</code>, and then the non-empty string evaluates to <code>true</code>.</p>

**Comment by arky** [comment]
<p>Thank you! That fixed it.</p>

**Comment by BrightRan** [comment, marked as answer]
<p><a class="mention" href="/arky">@arky</a>,</p>
<p>The problem is caused by the incorrect expressions you are using in the <strong>if</strong> conditional:</p>
<ol>
<li>You are using the incomplete expression (<strong>matrix.os</strong>) to access the value of <strong>os</strong> from the <a href="https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#contexts" rel="nofollow noopener"><strong>matrix context</strong></a>. The correct and complete expression should be “<strong>matrix.config.os</strong>”, and you can’t omit any node from the expression.</li>
</ol>
<pre><code class="lang-yaml">if: ${{ matrix.config.os == 'windows-latest' }}
</code></pre>
<ol start="2">
<li>Both sides of the logical operator (<strong>&amp;&amp;</strong> or <strong>||</strong>) must be complete conditional statements. You can’t omit any part from the conditional statements. The correct and complete conditional statements should be like as below.</li>
</ol>
<pre><code class="lang-yaml">if: ${{ matrix.config.os == 'ubuntu-latest' || matrix.config.os == 'macos-latest' }}
</code></pre>
<p>In summary, the correct configurations in your workflow should be like as below:</p>
<pre><code class="lang-yaml">jobs:
  build:
    runs-on: ${{ matrix.config.os }}
    strategy:
      matrix:
        config:
          - { os: macos-latest, arch: x64, name: macos-latest }
          - { os: ubuntu-latest, arch: x64, name: ubuntu-latest }
          - { os: windows-latest, arch: x64, name: windows-latest }
          - { os: windows-latest, arch: ia32, name: windows-latest-32 }
    steps:
      . . .

      - name: build with pyinstaller on gnu/Linux and MacOS
        if: ${{ matrix.config.os == 'ubuntu-latest' || matrix.config.os == 'macos-latest' }}
        . . .

      - name: build with pyinstaller on MS Windows
        if: ${{ matrix.config.os == 'windows-latest' }}
        . . .
</code></pre>

**Comment by arky** [comment]
<p>Thank you <a class="mention" href="/brightran">@brightran</a> <a class="mention" href="/airtower-luna">@airtower-luna</a></p>
<p>I ended up using the similar runner.os context instead of martix.config.os</p>
<p><code>if: runner.os == 'Linux' || runner.os == 'macOS'</code></p>
<p><code> if: runner.os == 'Windows'</code></p>
